atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java

```diff
@@ -166,26 +169,50 @@ public TransactionAndImmutableTsLock setupRunTaskWithConditionThrowOnConflict(Pr
         StartIdentifiedAtlasDbTransactionResponse transactionResponse
                 = timelockService.startIdentifiedAtlasDbTransaction();
         try {
-            LockToken immutableTsLock = transactionResponse.immutableTimestamp().getLock();
-            long immutableTs = transactionResponse.immutableTimestamp().getImmutableTimestamp();
-            recordImmutableTimestamp(immutableTs);
-
-            cleaner.punch(transactionResponse.startTimestampAndPartition().timestamp());
-            Supplier<Long> startTimestampSupplier = Suppliers.ofInstance(
-                    transactionResponse.startTimestampAndPartition().timestamp());
-
-            Transaction transaction = createTransaction(
-                    immutableTs,
-                    startTimestampSupplier,
-                    immutableTsLock,
-                    condition);
-            return TransactionAndImmutableTsLock.of(transaction, immutableTsLock);
+            return wrapResponse(condition, transactionResponse);
         } catch (Throwable e) {
             timelockService.tryUnlock(ImmutableSet.of(transactionResponse.immutableTimestamp().getLock()));
             throw Throwables.rewrapAndThrowUncheckedException(e);
         }
     }
 
+    @Override
+    public List<TransactionAndImmutableTsLock> setupRunTaskBatchWithConditionThrowOnConflict(
```

Comment:
Thoughts on returning something like TransactionAndImmutableTsLock, but that additionally has some sort of close method that users of this can call instead of using #finishRunTaskWithLockThrowOnConflict?

Reply:
I think I won't do that in this PR - we can maybe do that improvement in a separate task.